File: typescript/difference-of-squares/difference-of-squares.ts
function square(x: number): number {
  return Math.pow(x, 2);
}

function sumRange({
  start = 1,
  stop,
  unitFn = (x: number): number => x,
  sumFn = (x: number): number => x
}: {
  start?: number;
  stop: number;
  unitFn?: (x: number) => number;
  sumFn?: (x: number) => number;
}): number {
  let sum = 0;
  for (let i = start; i <= stop; i++) {
    sum += unitFn(i);
  }
  return sumFn(sum);
}

function squareOfSum(x: number): number {
  return sumRange({ stop: x, sumFn: square });
}

function sumOfSquares(x: number): number {
  return sumRange({ stop: x, unitFn: square });
}

export default class Squares {
  public squareOfSum: number;
  public sumOfSquares: number;
  public difference: number;

  constructor(x: number) {
    this.squareOfSum = squareOfSum(x);
    this.sumOfSquares = sumOfSquares(x);
    this.difference = this.squareOfSum - this.sumOfSquares;
  }
}

File: scala/difference-of-squares/src/main/scala/DifferenceOfSquares.scala
object DifferenceOfSquares {

  val sumOfSquares: Int => Int = n =>
    List
      .range(1, n + 1)
      .map(math.pow(_, 2))
      .sum
      .toInt

  val squareOfSum: Int => Int = (n) => {
    val sum = List.range(1, n + 1).sum
    math.pow(sum, 2).toInt
  }

  def differenceOfSquares(n: Int): Int =
    List(squareOfSum, sumOfSquares)
      .map(_(n))
      .reduce(_ - _)
}
